### PR TITLE
[pylint] Implement `too-many-public-methods` rule (PLR0904)

### PR DESCRIPTION
**Pull request by jelly**
Implement https://pylint.pycqa.org/en/latest/user_guide/messages/refactor/too-many-public-methods.html

Confusingly the rule page mentions a max of 7 while in practice it is 20. https://github.com/search?q=repo%3Apylint-dev%2Fpylint+max-public-methods&type=code

## Summary

Implement pylint's R0904

## Test Plan

Unit tests.

### CODE REVIEW

**Comment by jelly** [comment]
`python scripts/generate_mkdocs.py` succeeds locally for me, so I am not sure what's going on in CI.

File: crates/ruff/src/rules/pylint/rules/too_many_public_methods.rs [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    }
+}
+
+/// R0904
+pub(crate) fn too_many_public_methods(
+    checker: &mut Checker,
+    ast::StmtClassDef { name, body, .. }: &ast::StmtClassDef,
+    max_methods: usize,
+) {
+    let methods = body
```
Comment (konstin) [inline crates/ruff/src/rules/pylint/rules/too_many_public_methods.rs:108]: nit: destructuring in the function header is hard to read
```suggestion
    class_Def: &ast::StmtClassDef,
    max_methods: usize,
) {
    ast::StmtClassDef { name, body, .. } = class_def;
    let methods = body
```

File: crates/ruff/src/rules/pylint/rules/too_many_public_methods.rs [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        .iter()
+        .filter(|stmt| match stmt {
+            Stmt::FunctionDef(ast::StmtFunctionDef { name, .. }) => !name.starts_with('_'),
+            _ => true,
```
Comment (konstin) [inline crates/ruff/src/rules/pylint/rules/too_many_public_methods.rs:112]: wouldn't this count all fields? Relatedly, could you add some fields to the test case to make sure they aren't counted?

Comment (jelly) [reply crates/ruff/src/rules/pylint/rules/too_many_public_methods.rs:112]:
```
    0       │-too_many_public_methods.py:1:7: PLR0904 Too many public methods (10 > 7)
          0 │+too_many_public_methods.py:1:7: PLR0904 Too many public methods (11 > 7)
    1     1 │   |
    2     2 │ 1 | class Everything:
    3     3 │   |       ^^^^^^^^^^ PLR0904
    4       │-2 |     def __init__(self):
    5       │-3 |         pass
          4 │+2 |     foo = 1
```
Woops, this is indeed not functional.

Comment (jelly) [reply crates/ruff/src/rules/pylint/rules/too_many_public_methods.rs:112]: The problem is that I have `_ => true` which does not filter it.

**Review by konstin — APPROVED** [review]
lgtm, needs a review from @charliermarsh

**Comment by jelly** [comment]
@charliermarsh gentle ping.

File: crates/ruff/src/rules/pylint/rules/too_many_public_methods.rs [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+pub(crate) fn too_many_public_methods(
+    checker: &mut Checker,
+    class_def: &ast::StmtClassDef,
+    max_methods: usize,
+) {
+    let ast::StmtClassDef { name, body, .. } = class_def;
+    let methods = body
+        .iter()
+        .filter(|stmt| match stmt {
+            Stmt::FunctionDef(ast::StmtFunctionDef { name, .. }) => !name.starts_with('_'),
```
Comment (charliermarsh) [inline crates/ruff/src/rules/pylint/rules/too_many_public_methods.rs:112]: Can we use `method_visibility` here from `analyze::visibility`? It takes into account dunders and a few other cases.

File: crates/ruff/src/rules/pylint/rules/too_many_public_methods.rs [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    }
+}
+
+/// R0904
+pub(crate) fn too_many_public_methods(
+    checker: &mut Checker,
+    class_def: &ast::StmtClassDef,
+    max_methods: usize,
+) {
+    let ast::StmtClassDef { name, body, .. } = class_def;
```
Comment (charliermarsh) [inline crates/ruff/src/rules/pylint/rules/too_many_public_methods.rs:108]: Do you know if this rule should be applied to private classes? What about to classes in private modules? Do you mind checking how Pylint treats those?

Comment (jelly) [reply crates/ruff/src/rules/pylint/rules/too_many_public_methods.rs:108]: What would a "private module" be in Python? `_module.py` ? As far as I can see [the rule](https://github.com/pylint-dev/pylint/blob/main/pylint/checkers/design_analysis.py#L473) counts the methods except dunders. Well I assume it also ignores `def _foo()` as that would be `private` by convention in Python.

File: crates/ruff/src/rules/pylint/rules/too_many_public_methods.rs (hunk 3 above)
Comment (charliermarsh) [reply crates/ruff/src/rules/pylint/rules/too_many_public_methods.rs:112]: I think it's okay to make it `pub` and use it here.

File: crates/ruff/src/rules/pylint/rules/too_many_public_methods.rs (hunk 4 above)
Comment (charliermarsh) [reply crates/ruff/src/rules/pylint/rules/too_many_public_methods.rs:108]: Yeah, something like `_module.py`, which would be considered private based on pydocstyle's definitions of public vs. private visibility for functions.

File: crates/ruff/src/rules/pylint/rules/too_many_public_methods.rs (hunk 3 above)
Comment (jelly) [reply crates/ruff/src/rules/pylint/rules/too_many_public_methods.rs]: Applied, the changes. Please have a look.

**Pull request merged**